packages/xod-project/src/patch.js

```diff
@@ -1366,3 +1380,140 @@ export const validateAbstractPatch = def(
     Either.of
   )
 );
+
+export const getPatchSignature = def(
+  'getPatchSignature :: Patch -> PatchSignature',
+  R.compose(
+    R.map(
+      R.compose(
+        R.fromPairs,
+        R.map(R.converge(R.pair, [Pin.getPinOrder, Pin.getPinType]))
+      )
+    ),
+    R.groupBy(Pin.getPinDirection),
+    listPins
+  )
+);
+
+// checks that a concrete patch conforms to concrete patch specification
+export const checkSpecializationMatchesAbstraction = def(
+  'checkSpecializationMatchesAbstraction :: Patch -> Patch -> Either Error Patch',
+  (abstractPatch, specializationPatch) => {
+    if (isAbstractPatch(specializationPatch)) {
+      return fail('SPECIALIZATION_PATCH_CANT_BE_ABSTRACT', {});
+    }
+
+    if (
+      !R.equals(
+        getArityStepFromPatch(abstractPatch),
+        getArityStepFromPatch(specializationPatch)
+      )
+    ) {
+      return fail('SPECIALIZATION_PATCH_MUST_HAVE_SAME_ARITY_LEVEL', {});
```

Comment:
I think it's better to provide some additional information about the error in the payload (arity and patchPaths?)
This comment relates to some other errors in this function.